### PR TITLE
fix multiple varuna bugs; run varuna with gpt example

### PR DESCRIPTION
**Pull request by XianzheMa**

### CODE REVIEW

File: examples/gpt/README.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+## troubleshooting
+- If there are weird errors during model `.forward()` execution,
+try to clear all tmp files with prefix `_tmp_` in `examples/gpt/` directory. They are cache storing structure information
+about the model. If you change the model structure, you need to clear them.
```
Comment (fotstrt) [inline examples/gpt/README.md:4]: Is this also true in the case that a worker gets a different part of the model upon reconfiguration? (i.e. the model is split differently than before)

File: install_script.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+set -e
+
+git clone https://github.com/XianzheMa/varuna.git -b xianzhe/redact_varuna
```
Comment (fotstrt) [inline install_script.sh:5]: Is the `xianzhe/redact_varuna` branch active? I don't see it

File: varuna/launcher.py [hunk 3]
```diff
@@ -122,7 +122,7 @@ def parse_args():
                         help="The total number of nodes.")
     parser.add_argument("--node_rank", type=int, default=0,
                         help="Rank of node amongst servers.")
-    parser.add_argument("--nstages", type=int, required = True,
+    parser.add_argument("--nstages", type=int, default=None,
```
Comment (fotstrt) [inline varuna/launcher.py:125]: Why this change?

File: varuna/auto_config.py [hunk 4]
```diff
@@ -133,7 +137,7 @@ def get_min(self):
     def read_profile(self, profile_folfder, autofill_missing_compute=False):
         
         self.compute_profile = []
-        self.comm_profile = []
+        self.comm_profile = None
```
Comment (fotstrt) [inline varuna/auto_config.py:140]: Why this change?

File: varuna/profiler.py [hunk 5]
```diff
@@ -55,12 +55,14 @@ def receiver(recv_rank, recv_shape, recv_times, dtype):
         if recv_handles.qsize() > 4:
             handle, start_time = recv_handles.get()
             handle.wait()
-            recv_times.append(time.time() - start_time)
+            # convert seconds to microseconds
```
Comment (fotstrt) [inline varuna/profiler.py:58]: Why are you converting to microseconds here?

File: examples/gpt/model/gpt_modules.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
+import math
```
Comment (fotstrt) [inline examples/gpt/model/gpt_modules.py:1]: I would recommend, for files like this, (and also like `pretrain_corpus.py` and `tokenizer.py` I guess), to put a comment at the start of the file, declaring the source of the file (which repo is it from)

**Review by fotstrt — COMMENTED** [review]
Looks great! I added some comments because I am not sure I understand why you are making some changes.

File: examples/gpt/README.md (hunk 1 above)
Comment (XianzheMa) [reply examples/gpt/README.md:4]: Yes. but all these `_tmp_*` files store the information about the complete model, not any specific partition. 
So a worker might get a different part of the model upon reconfiguration but it won't cause any issue with the `_tmp_*` files.

File: install_script.sh (hunk 2 above)
Comment (XianzheMa) [reply install_script.sh:5]: Good catch thanks! I will fix it

Comment (XianzheMa) [reply install_script.sh:5]: I will fix it in next PR

File: varuna/launcher.py (hunk 3 above)
Comment (XianzheMa) [reply varuna/launcher.py:125]: `launcher.py` is the entry point on each worker node to launch the training process.
When `--nstages` is left unspecified, it means we want to auto-configure it by varuna.
I believe this is a bug on original Varuna, because in their original repo, it just uses `if args.nstages is None` to decide if auto-config is needed. See [here](https://github.com/microsoft/varuna/blob/a2145cc30b976a7917829a55e2a866f0ca4c54e5/varuna/launcher.py#L21)

File: varuna/auto_config.py (hunk 4 above)
Comment (XianzheMa) [reply varuna/auto_config.py:140]: because self.comm_profile is not intended to be a list, rather to be an object. See https://github.com/microsoft/varuna/blob/a2145cc30b976a7917829a55e2a866f0ca4c54e5/varuna/auto_config.py#L149

File: examples/gpt/model/gpt_modules.py (hunk 6 above)
Comment (XianzheMa) [reply examples/gpt/model/gpt_modules.py:1]: I will in my next PR! Thanks

File: varuna/profiler.py (hunk 5 above)
Comment (XianzheMa) [reply varuna/profiler.py:58]: This is a bug in varuna.

So first, one profiling pass can only profile up to `world_size` cutpoints, so we might need multiple profiling passes to profile the entire model.

Some information is profiled every profiling pass (e.g. this communication latency), and the program simply stores the average of all, and it should be saved in microseconds.

The original logic is:
1st pass:
1. get communication latency 0.001s. 
2. Store it in microseconds  (0.001s * 1000000 = 1000 μs) in the profile file.

2nd pass:
1. get the communication latency 0.0011s.
2. Read last average from file: 1000 μs.
3. Simply calculate the average of the two figures **ignoring the unit**: (1000 + 0.0011)/2 = 500.00055
4. Store it in microseconds in the profile file. The number stored is 500.00055 * 1000

As you can see there is unit conversion bug. Now I pre-convert each sample to microseconds, and in the final averaging phase, remove the conversion to microseconds. Then the result is correct.
